### ISSUE TITLE
Support specifying number of iterations in dataset evaluation

### ISSUE THREAD

**Issue by wangmn93** [microsoft/LightGBM#4210]
I want to evaluate a constructed dataset using only one tree.  booster.predict use the raw data and it is slow.  Is there any way to do this?

import numpy as np
from lightgbm import train, Booster

data = np.random(0,1,(100,10))
label = np.random(0,1,100)
dataset = Dataset(data=data, label=label)
booster =  train(dataset, num_boost_round=10)
tree0 = Booster(model_string=booster.model_to_string(start_iteration=0, num_iteration=1))

tree0.predict(data) # this is slow

tree0 = Booster(model_str=booster.model_to_string(start_iteration=0, num_iteration=1), train_set=dataset)

tree0._Booster__inner_predict(0) # it always output zeros, since the init score is zero

**Comment by shiyu1994** [comment]
@wangmn93 Thanks for using LightGBM. If you want to evaluate a constructed model, you have to make it as either a training data or validation data, and use it in the training process.
Once the model is trained, currently we don't have any support to use the trained model to evaluate a constructed `Dataset`.
Also, I don't think prediction with a constructed `Dataset` (if we can do this) should be faster than with raw data. How slow it is to predict with raw data in your case?

**Comment by wangmn93** [comment]
@shiyu1994 I have a matrix of 1000000 x 1000, it takes 8.7ms using constructed dataset and 6.85s using booster.predict.  Besides the prediction time, it is preferred to use constructed dataset since the raw data is big and loading data takes a lot of time.

**Comment by StrikerRUS** [comment]
> Once the model is trained, currently we don't have any support to use the trained model to evaluate a constructed `Dataset`.

I think `eval(_train/_valid)` functions can be used to score a constructed Dataset.
https://github.com/microsoft/LightGBM/issues/3949#issuecomment-808810679
https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Booster.html#lightgbm.Booster.eval

**Comment by StrikerRUS** [comment]
@shiyu1994 Do you think we can add a feature request for including `start_iteration` and `num_iteration` (just like in the `predict()` method) into `eval(_train/_valid)` methods?

**Comment by shiyu1994** [comment]
@StrikerRUS Sure. That will be valuable.

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute to this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.
